ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+  | EpLambda EpType EpType
+  | EpMap EpType EpType
+  | EpBigMap EpType EpType
+  | EpBytes
+  | EpAddress
+  | EpKey
+  | EpChainId
+  | EpMutez
+  | EpBool
+  | EpSet EpType
+  | EpKeyHash
+  | EpTimestamp
+  | EpList EpType
+  | EpOption EpType
+  | EpPair AnnotatedField AnnotatedField
+  | EpOr AnnotatedField AnnotatedField
+  deriving stock (Show, Generic)
+
+type AnnotatedField
+ = (FieldAnn, EpType)
+
+-- An entrypoint extracted from parameter with entrypoint name
+-- and the type of the parameter it wraps.
+data Entrypoint = Entrypoint
+  { epName :: Text
+  , epType :: EpType
+  } deriving stock (Show, Generic)
+
+toEpType :: Notes t -> EpType
+toEpType (NTString _) = EpString
+toEpType (NTInt _) = EpInt
+toEpType (NTNat _) = EpNat
+toEpType (NTUnit _) = EpUnit
+toEpType (NTSignature _) = EpSignature
+toEpType (NTOption _ a) = EpOption (toEpType a)
+toEpType (NTOr _ n1 n2 f1 f2) = EpOr (n1, toEpType f1) (n2, toEpType f2)
+toEpType (NTPair _ n1 n2 f1 f2) = EpPair (n1, toEpType f1) (n2, toEpType f2)
+toEpType (NTKey _) = EpKey
+toEpType (NTChainId _) = EpChainId
+toEpType (NTList _ a) = EpList (toEpType a)
+toEpType (NTSet _ a) = EpSet (toEpType a)
+toEpType (NTOperation _) = EpOperation
+toEpType (NTMap _ f1 f2) = EpMap (toEpType f1) (toEpType f2)
+toEpType (NTBigMap _ f1 f2) = EpBigMap (toEpType f1) (toEpType f2)
+toEpType (NTContract _ a) = EpContract (toEpType a)
+toEpType (NTLambda _ f1 f2) = EpLambda (toEpType f1) (toEpType f2)
+toEpType (NTBytes _) = EpBytes
+toEpType (NTMutez _) = EpMutez
+toEpType (NTKeyHash _) = EpKeyHash
+toEpType (NTTimestamp _) = EpTimestamp
+toEpType (NTAddress _) = EpAddress
+toEpType (NTBool _) = EpBool
+
+-- looks up the entrypoints in the Notes t for the whole parameter.
+-- It might be possible to do this without re-implementing logic in
+-- the `mkEntrypointsMap` from morley.
+lookupEntryPointsAndTypes :: Notes t -> [Entrypoint]
+lookupEntryPointsAndTypes (NTOr _ fa1 fa2 f1 f2) =
+  let
+    leftEntrypoint = case extractAnn fa1 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f1)) : (lookupEntryPointsAndTypes f1)
+      Nothing -> lookupEntryPointsAndTypes f1
+    rightEntrypoint = case extractAnn fa2 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f2)) : (lookupEntryPointsAndTypes f2)
+      Nothing -> lookupEntryPointsAndTypes f2
+  in leftEntrypoint <> rightEntrypoint
+lookupEntryPointsAndTypes _ = []
+
+-- Check if the annotation in this annotation is among the
+-- ones we are not interested in, or empty. If so return a Nothing. Else
+-- wrap the annotation in a Just before returning it.
+extractAnn :: Annotation tag -> Maybe Text
+extractAnn ann_ =
+  case (ann_ == noAnn) || (isJust $ find (== ann_) ignoreEntypoints) of
+    True -> Nothing
+    _ -> Just $ unAnnotation ann_
+
+-- Some entrypoints we don't want to call directly.
+ignoreEntypoints :: [Annotation tag]
+ignoreEntypoints = [ann "running", ann "transfer_contract_tokens", ann "xtzForbidden", ann "call_FA2", ann "migratable"]
```

Comment:
Sorry, but why do we want to ignore those entrypoints?

Reply:
These are not "real" entrypoints, and they have real entrypoints deeper down the tree. 

If you look at our parameter [here](https://github.com/tqtezos/baseDAO/blob/master/ligo/haskell/src/Ligo/BaseDAO/Types.hs#L116), you can see that it is a nested type, instead of a single, flat sum type. And this is so that we can remove duplication in terms of contract code for some checks that otherwise would have to be repeated for each separate entrypoints. Because of this nesting, we can do such checks once, on the top level branch. In the ligo contract, we have manually constructuring these stuctures so that these entrypoints wont show in the actual parameter.

That being said, there is indeed one bug here, which is, `transfer_contract_tokens` should not be ignored here, so I ll fix it.

```
parameter
  (or (option %startup (pair string (option bytes)))
      (or (or (pair %callCustom string bytes)
              (or (or (or (or (unit %accept_ownership)
                              (pair %burn (address %from_) (pair (nat %token_id) (nat %amount))))
                          (or (or %call_FA2
                                 (or (pair %balance_of
                                        (list %requests (pair (address %owner) (nat %token_id)))
                                        (contract %callback
                                           (list (pair (pair %request (address %owner) (nat %token_id)) (nat %balance)))))
                                     (list %transfer
                                        (pair (address %from_)
                                              (list %txs (pair (address %to_) (pair (nat %token_id) (nat %amount)))))))
                                 (list %update_operators
                                    (or (pair %add_operator (address %owner) (pair (address %operator) (nat %token_id)))
                                        (pair %remove_operator (address %owner) (pair (address %operator) (nat %token_id))))))
                              (unit %confirm_migration)))
                      (or (or (bytes %drop_proposal) (nat %flush))
                          (or (pair %getVotePermitCounter (unit %param) (contract %callback nat))
                              (pair %get_total_supply (nat %token_id) (contract %callback nat)))))
                  (or (or (or (address %migrate) (pair %mint (address %to_) (pair (nat %token_id) (nat %amount))))
                          (or (pair %propose (nat %frozen_token) (map %proposal_metadata string bytes))
                              (nat %set_quorum_threshold)))
                      (or (or (nat %set_voting_period) (address %transfer_ownership))
                          (list %vote
                             (pair (pair %argument (bytes %proposal_key) (pair (bool %vote_type) (nat %vote_amount)))
                                   (option %permit (pair (key %key) (signature %signature)))))))))
          (pair %transfer_contract_tokens
             (address %contract_address)
             (list %params
                (pair (address %from_)
                      (list %txs (pair (address %to_) (pair (nat %token_id) (nat %amount))))))))) ;
```

Reply:
Ah, I see, thanks for clarifying!

Reply:
Have removed the 'transfer_contract_tokens' entrypoint from list of ignored ones.